newrelic-agent/src/main/java/com/newrelic/agent/stats/dimensional/MeterService.java

```diff
@@ -0,0 +1,314 @@
+        metricData.forEach(md -> {
+            switch (md.getType()) {
+                case LONG_SUM:
+                    md.getLongSumData().getPoints().forEach(pointData -> {
+                        final Map<String, Object> intrinsics = getIntrinsics(md.getName(), pointData);
+                        intrinsics.put("metric.count", pointData.getValue());
+                        events.add(createEvent(pointData, intrinsics));
+                    });
+                    break;
+                case HISTOGRAM:
+                    md.getHistogramData().getPoints().forEach(pointData -> {
+                        final Map<String, Object> intrinsics = getIntrinsics(md.getName(), pointData);
+                        intrinsics.put("metric.summary", toSummary(pointData));
+                        events.add(createEvent(pointData, intrinsics));
+                    });
+            }
+        });
+        return events;
+    }
+
+    private static CustomInsightsEvent createEvent(PointData pointData, Map<String, Object> intrinsics) {
+        final long timestamp = TimeUnit.NANOSECONDS.toMillis(pointData.getStartEpochNanos());
+        return new CustomInsightsEvent("Metric", timestamp, toMap(pointData.getAttributes()), intrinsics, DistributedTraceServiceImpl.nextTruncatedFloat());
+    }
+
+    private static List<Number> toSummary(HistogramPointData pointData) {
+        return ImmutableList.of(pointData.getCount(), pointData.getSum(), pointData.getMin(), pointData.getMax());
+    }
+
+    static Map<String, Object> getIntrinsics(String metricName, PointData pointData) {
+        final Map<String, Object> intrinsics = new HashMap<>();
+        intrinsics.put("metric.name", metricName);
+        final long durationMs = TimeUnit.NANOSECONDS.toMillis(pointData.getEpochNanos() - pointData.getStartEpochNanos());
+        intrinsics.put("duration.ms", durationMs);
+        return intrinsics;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    public void addHarvestableToService(String appName) {
+        if (!enabled) {
+            return;
+        }
+        final String defaultAppName = ServiceFactory.getRPMService().getApplicationName();
+        if (appName.equals(defaultAppName)) {
+            final Harvestable harvestable = new Harvestable(this, appName) {
+                @Override
+                public String getEndpointMethodName() {
+                    return CollectorMethods.DIMENSIONAL_METRIC_DATA;
+                }
+
+                @Override
+                public int getMaxSamplesStored() {
+                    return ServiceFactory.getConfigService().getDefaultAgentConfig().getInsightsConfig().getMaxSamplesStored();
+                }
+            };
+            ServiceFactory.getHarvestService().addHarvestable(harvestable);
+            closeables.add(() -> ServiceFactory.getHarvestService().removeHarvestable(harvestable));
+        }
+    }
+
+    interface Closeable {
```

Comment:
#nit: Use `java.lang.AutoCloseable` instead. It throws Exception, but its easy to catch that and log / ignore that while iterating through.